storage/fanout.go

```diff
@@ -289,99 +319,82 @@ func (q *mergeGenericQuerier) Select(sortSeries bool, hints *SelectHints, matche
 	for r := range seriesSetChan {
 		seriesSets = append(seriesSets, r)
 	}
-	return &lazySeriesSet{create: create(seriesSets, q.mergeFn)}
+	return &lazyGenericSeriesSet{init: func() (genericSeriesSet, bool) {
+		s := newGenericMergeSeriesSet(seriesSets, q.mergeFn)
+		return s, s.Next()
+	}}
 }
 
-func create(seriesSets []genericSeriesSet, mergeFunc genericSeriesMergeFunc) func() (genericSeriesSet, bool) {
-	// Returned function gets called with the first call to Next().
-	return func() (genericSeriesSet, bool) {
-		if len(seriesSets) == 1 {
-			return seriesSets[0], seriesSets[0].Next()
-		}
-		var h genericSeriesSetHeap
-		for _, set := range seriesSets {
-			if set == nil {
-				continue
-			}
-			if set.Next() {
-				heap.Push(&h, set)
-				continue
-			}
-			// When primary fails ignore results from secondaries.
-			// Only the primary querier returns error.
-			if err := set.Err(); err != nil {
-				return errorOnlySeriesSet{err}, false
-			}
-		}
-		set := &genericMergeSeriesSet{
-			mergeFunc: mergeFunc,
-			sets:      seriesSets,
-			heap:      h,
-		}
-		return set, set.Next()
-	}
+type labelGenericQueriers []genericQuerier
+
+func (l labelGenericQueriers) Len() int               { return len(l) }
+func (l labelGenericQueriers) Get(i int) LabelQuerier { return l[i] }
+func (l labelGenericQueriers) SplitByHalf() (labelGenericQueriers, labelGenericQueriers) {
+	i := len(l) / 2
+	return l[:i], l[i:]
 }
 
 // LabelValues returns all potential values for a label name.
 func (q *mergeGenericQuerier) LabelValues(name string) ([]string, Warnings, error) {
-	var (
-		results  [][]string
-		warnings Warnings
-	)
-	for _, querier := range q.queriers {
-		values, wrn, err := querier.LabelValues(name)
-		if wrn != nil {
-			// TODO(bwplotka): We could potentially wrap warnings.
-			warnings = append(warnings, wrn...)
-		}
-		if err != nil {
-			return nil, nil, errors.Wrapf(err, "LabelValues() from Querier for label %s", name)
+	res, ws, err := q.lvals(q.queriers, name)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "LabelValues() from merge generic querier for label %s", name)
+	}
+	return res, ws, nil
+}
+
+func (q *mergeGenericQuerier) lvals(lq labelGenericQueriers, n string) ([]string, Warnings, error) {
+	if lq.Len() == 0 {
+		return []string{}, nil, nil
+	}
+	if lq.Len() == 1 {
+		return lq.Get(0).LabelValues(n)
+	}
+	a, b := lq.SplitByHalf()
+
+	var ws Warnings
+	s1, w, err := q.lvals(a, n)
+	ws = append(ws, w...)
+	if err != nil {
+		return nil, ws, err
+	}
+	s2, ws, err := q.lvals(b, n)
+	ws = append(ws, w...)
+	if err != nil {
+		return nil, ws, err
+	}
+	return mergeStrings(s1, s2), ws, nil
+}
+
+func mergeStrings(a, b []string) []string {
```

Comment:
Unexported but non-trivial function. Would be great to have a doc comment describing what it does.